File: src-tauri/src/icloud.rs
// The iCloud Drive source, iOS only.
//
// Unlike storage.rs's local folder, there is no folder to pick: an app's own
// ubiquity container is fixed by its identifier (iCloud.review.dev.app,
// registered against this project's Apple Developer team - see
// reviewer_iOS.entitlements) and NSFileManager hands back its URL directly.
// Everything past that root - listing, reading, writing, removing - is
// storage.rs's own storage_list/storage_read/storage_write/storage_remove,
// unchanged: they already take an arbitrary root and don't care how it was
// chosen, so this file's only job is resolving the one this source has.

use objc2_foundation::{NSFileManager, NSString};

// Matches com.apple.developer.icloud-container-identifiers in
// reviewer_iOS.entitlements - the two have to agree, or the container this
// asks for is one the app was never granted.
const CONTAINER: &str = "iCloud.review.dev.app";

/// The Documents folder inside this app's iCloud container, creating it if
/// this is the first time - iCloud does not create it on its own, and a
/// listing against a folder that has never existed reads the same as a
/// listing against an empty one, but a write into it needs somewhere real to
/// land first.
///
/// `Ok(None)` rather than an error when iCloud itself is not available -
/// signed out of iCloud, or the device has it turned off for this app -
/// which is a reader's ordinary account state, not a fault. The frontend's
/// `ready()` already tells them what to do about a source with no root, the
/// same sentence a local folder that was never chosen gets.
#[tauri::command]
pub fn icloud_root() -> Result<Option<String>, String> {
    let manager = NSFileManager::defaultManager();
    let identifier = NSString::from_str(CONTAINER);

    let Some(container) = manager.URLForUbiquityContainerIdentifier(Some(&identifier)) else {
        return Ok(None);
    };

    let documents_name = NSString::from_str("Documents");
    let documents = container
        .URLByAppendingPathComponent(&documents_name)
        .ok_or_else(|| "could not resolve the Documents folder inside the iCloud container".to_string())?;

    let path = documents
        .path()
        .ok_or_else(|| "the iCloud container's Documents folder has no filesystem path".to_string())?
        .to_string();

    std::fs::create_dir_all(&path).map_err(|error| error.to_string())?;

    Ok(Some(path))
}

File: web/src/adapters/icloud.js
// A reader backed by this app's own iCloud container, on iOS.
//
// Unlike TauriAdapter's local folder, there is nothing to pick: the container
// is fixed by the app's own identifier (see src-tauri/src/icloud.rs and
// reviewer_iOS.entitlements), so attaching this source is asking Rust for its
// path rather than opening a dialog for the reader to choose one. Everything
// past that root is identical to a local folder's - the same storage_list/
// storage_read/storage_write/storage_remove commands, which do not know or
// care that a root came from iCloud instead of a picker - so this extends
// TauriAdapter rather than repeating its list/read/write/remove/media.

import { TauriAdapter, inTauriIOS } from "./tauri.js";

function core() {
  const api = globalThis.__TAURI__;

  return api?.core?.invoke ? api.core : null;
}

/**
 * Ask Rust for this app's iCloud container path, creating it if this is the
 * first time.
 *
 * @returns {Promise<string|null>} the path, or null when iCloud itself is not
 *   available - signed out of iCloud, or turned off for this app
 * @throws {Error} if asking failed outright, which is not the same as iCloud
 *   simply not being signed in
 */
export async function icloudRoot() {
  const api = core();

  if (!api) throw new Error("the iOS app is not running");

  try {
    return (await api.invoke("icloud_root")) ?? null;
  } catch (failure) {
    throw failure instanceof Error ? failure : new Error(String(failure));
  }
}

/**
 * Why this source cannot be used here, if it cannot.
 *
 * @returns {{reason: string, hint: string}} usable when both are ""
 */
export function unavailability() {
  if (inTauriIOS()) return { reason: "", hint: "" };

  return { reason: "needs the iOS app", hint: "" };
}

export class ICloudAdapter extends TauriAdapter {
  static type = "icloud";
  static label = "iCloud Drive";

  // Nothing to ask - see icloudRoot() above for what fills this in instead.
  static fields = [];

  static precise = false;

  async ready() {
    if (!inTauriIOS()) {
      return { ok: false, reason: "this reader needs the iOS app" };
    }

    // Re-resolved rather than trusted from config(): a reader can sign out of
    // iCloud, or turn it off for this app, between sessions, and a root that
    // was real yesterday is not proof of anything about right now.
    const root = await icloudRoot();

    if (!root) {
      return { ok: false, reason: "iCloud isn't available - check that you're signed in" };
    }

    this._root = root;

    return { ok: true, reason: "" };
  }

  config() {
    // No root kept here: it comes from icloudRoot() fresh every ready(), and
    // a stale one persisted from a previous device or account would be a
    // path this app never actually resolved, not a shortcut past resolving
    // it again.
    return { type: ICloudAdapter.type, label: this._label };
  }

  describe() {
    return this._label || "This app's iCloud Drive";
  }
}
